# Chapter 8 - Interest Rates and Bond Valuation

Corporate Finance 9th International Edition

Ross, Westerfield, Jaffe

In [26]:
import pandas as pd
import numpy_financial as npf

# Lets define some function so we can reuse them later.

def zero_coupon(face, ytm, years, compound=1):
    
    return face / (1 + (ytm / compound))**(years*compound)

def bond_value(face, ytm, periods, coupon, compound=1):
    
    bv = coupon * ((1 - (1/(1+(ytm/compound)))**(periods*compound)) / (ytm/compound)) + \
        face * (1 / (1 + (ytm/compound))**(periods*compound)) 
    
    return bv

def realRate(rate, inflation):
    """Returns Real Rate (Nominal Rate Adjusted for Inflation)"""
    inflation = inflation / 100
    real = ((1 + rate) / (1 + inflation)) - 1
    return real


**1. Valuing Bonds** - What is the price of a 10-year, zero coupon bond paying 1,000 at maturity if the YTM is:

* a. 5 percent?
* b. 10 percent?
* c. 15 percent?

In [9]:
#Given
years = 10

amount = 1000
ytms = [.05, .10, .15]

prices = []
columns = []

for yields in ytms:
    #
    price = f'{zero_coupon(amount, yields, years, compound=2):,.2f}'
    col = f'{yields*100}%'
    prices.append(price)
    columns.append(col)

df = pd.DataFrame([prices], columns=columns)
df


5.0%   10.0%   15.0%
0  610.27  376.89  235.41

**Comment**

Even though there are no coupon payments, the periods are semiannual to stay consistent with coupon bond payments.  That is why we specify compound=2.

**2. Valuing Bonds** - Microhard has issued a bond with the following characteristics:

* Par: 1,000
* Time to maturity: 25 years
* Coupon rate: 7 percent
* Semiannual payments

Calculate the price of this bond if the YTM is:
* a. 7 percent
* b. 9 percent
* c. 5 percent

In [15]:
# Given
face = 1000
periods = 25
coupon_rate = .07
compound = 2 # semiannual
coupon = (face * coupon_rate)/compound

prices = []
columns = []
ytms = [.07, .09, .05]

for y in ytms:
    
    price = f'{bond_value(face, y, periods, coupon, compound=2):,.2f}'
    col = f'{y*100:.0f}%'
    prices.append(price)
    columns.append(col)

df = pd.DataFrame([prices], columns=columns)
df

7%      9%        5%
0  1,000.00  802.38  1,283.62

**Comment**

* a. `1,000` - When the YTM and the coupon rate are equal, the bond will sell at par.
* b. `802.38` - When the YTM is greater than the coupon rate, the bond will sell at a discount.
* c. `1,283.62` - When the YTM is less than the coupon rate, the bond will sell at a premium.

**6. Bond Yields** - A Japanese company has a bond outstanding that sells for 87 percent of it s 100,000 par value.  The bond has a coupon rate of 5.4 percent paid annually and matures in 21 years.  What is the yield to maturity of this bond?

In [18]:
# Given information
face_value = 100000
bond_value = face_value * 0.87  # Assuming the bond value is 87% of the face value
coupon_rate = 0.054
periods = 21
compounding = 1  # Assuming yearly compounding

# Calculate coupon payments
coupon_payment = face_value * coupon_rate

# Generate cash flows
cash_flows = [-bond_value] + [coupon_payment] * periods + [coupon_payment + face_value]

# Calculate yield to maturity (YTM)
ytm = npf.irr(cash_flows)

answer = f'Yield to maturity (YTM) of the bond: {ytm*100:.2f}'
answer

'Yield to maturity (YTM) of the bond: 6.53'

**7. Calculating Real Rates of Return** - If Treasury bills are currently paying 5 percent and the inflation rate is 3.9 percen, what is the approximate real rate of interest?  The exact real rate?

In [41]:
#Given
rate = .05
inflation = .039

approximate = treasury_bill - inflation
fish = ((1 + rate)/(1+inflation)) - 1

answer = f'The aproximate real rate of interest is {approximate*100:.2f}%, while \
the exact real rate is {fish*100:.2f}%.'
answer

'The aproximate real rate of interest is 1.10%, while the exact real rate is 1.06%.'

**8. Inflation and Nominal Returns** - Suppose the real rate is 2.5 percent and the inflation rate is 4.7 percent.  What rate would you expect to see on a Treasury bill?

In [40]:
#given 
real_rate = .025
inflation = .047

fish = ((1 + real_rate) * (1 + inflation)) - 1

answer = f'I would expect to see a rate of {fish*100:.2f}% on a treasury bill.'
answer

'I would expect to see a rate of 7.32% on a treasury bill.'